smolyay/basis.py

```diff
@@ -403,29 +482,393 @@ def _derivative(self, x):
+        """numpy.ndarray: the starting index of each level."""
+        return self._start_level
+
+    @property
+    def end_level(self):
+        """numpy.ndarray: the ending index of each level."""
+        return self._end_level
+
+    def level(self, index):
+        """list of :class:BasisFunction: Functions in a level"""
+        return self.basis_functions[self.start_level[index] : self.end_level[index]]
+
+
+class NestedCustomBasisFunctionSet(NestedBasisFunctionSet):
+    """Set of nested basis functions and sample points.
+
+    Parameters
+    ----------
+    basis_functions : list
+        Basis functions in set.
+
+    num_per_level : list
+        number of unique functions per level.
+
+    Raises
+    ------
+    IndexError
+        Must have at least one basis function
+    ValueError
+        Basis functions must have the same domain.
+    IndexError
+        number of basis function does not match functions in each level.
+    """
+
+    def __init__(self, basis_functions, num_per_level):
+        if sum(num_per_level) != len(basis_functions):
+            raise IndexError(
+                str(sum(num_per_level))
+                + " total functions in levels, "
+                + str(len(basis_functions))
+                + " functions given."
+            )
+        if len(basis_functions) == 0:
+            raise IndexError("Must have at least one term.")
+        domain = basis_functions[0].domain
+        if any(not numpy.array_equal(domain, b.domain) for b in basis_functions[1:]):
+            raise ValueError("Basis functions must have the same domain.")
+        self._basis_functions = basis_functions
+        self._num_per_level = numpy.array(num_per_level, dtype=int)
+        self._end_level = numpy.cumsum(self._num_per_level)
+        self._start_level = self._end_level - self._num_per_level
+
+
+class MutableNestedBasisFunctionSet(NestedBasisFunctionSet):
+    """Set of basis functions where number of functions is mutable.
+
+    Parameters
+    ----------
+    num_terms : int
+        the number of terms in the set.
+    """
+
+    def __init__(self, num_levels):
+        self._num_levels = None
+        self._basis_functions = None
+        self._start_level = None
+        self._end_level = None
+        self._num_per_level = None
+
+        self.num_levels = num_levels
+
+    @property
+    def num_levels(self):
+        """int: number of levels."""
+        return self._num_levels
+
+    @num_levels.setter
+    def num_levels(self, value):
+        num_levels = int(value)
+        if num_levels <= 0:
+            raise ValueError("Must have at least one level.")
+        if num_levels != self._num_levels:
+            self._create_levels(num_levels)
+            self._create()
+            self._num_levels = num_levels
+
+    @abc.abstractmethod
+    def _create(self):
+        """Create the basis functions in the set."""
+        pass
+
+
+class NestedClenshawCurtisBasisFunctionSet(
```

Comment:
Have this class derive directly from the `NestedBasisFunctionSet` class, then generate the list of lists in the constructor before calling the super. This is where you can use the method you put in `_growth`. Something like:
```
def __init__(self, num_levels):
    num_per_level = _growth.clenshaw_curtis_exponential_growth(num_levels)
    num_terms = np.sum(num_per_level)
    basis_functions = [ChebyshevFirstKind(n) for n in range(num_terms)]
    super().__init__(basis_functions, num_per_level)
```